File: frontend/src/testing/orchestrator/TestOrchestrator.ts
/**
 * TestOrchestrator - Orchestrates test execution with fail-fast behavior
 */

import { BaseCommand, TestContext, TestResult } from '../commands/BaseCommand';
import { PipelineController, PipelineConfig } from '../utils/PipelineController';
import { issueTracker } from '../utils/IssueTracker';

export interface OrchestratorConfig {
  baseUrl: string;
  timeout: number;
  retries: number;
  failFast: boolean;
  parallel: boolean;
  maxConcurrency: number;
  reportPath: string;
}

export interface TestSuite {
  name: string;
  commands: BaseCommand[];
  timeout?: number;
  retries?: number;
  failFast?: boolean;
}

export interface TestReport {
  timestamp: string;
  totalTests: number;
  passedTests: number;
  failedTests: number;
  skippedTests: number;
  totalDuration: number;
  suites: TestSuiteResult[];
  summary: {
    successRate: number;
    averageDuration: number;
    fastestTest: string;
    slowestTest: string;
  };
  pipelineResult?: any;
  issues?: any;
}

export interface TestSuiteResult {
  name: string;
  success: boolean;
  duration: number;
  tests: TestResult[];
  errors: string[];
}

export class TestOrchestrator {
  private config: OrchestratorConfig;
  private context: TestContext;
  private pipelineController: PipelineController;

  constructor(config: OrchestratorConfig) {
    this.config = config;
    this.context = {
      baseUrl: config.baseUrl,
      timeout: config.timeout,
      retries: config.retries,
      failFast: config.failFast,
      page: undefined,
      browser: undefined
    };
    
    // Initialize pipeline controller
    const pipelineConfig: PipelineConfig = {
      failFast: config.failFast,
      maxRetries: config.retries,
      retryDelay: 2000, // 2 seconds
      autoAddressIssues: true,
      criticalIssueThreshold: 3,
      stopOnCriticalIssues: true
    };
    
    this.pipelineController = new PipelineController(pipelineConfig);
  }

  async runSuites(suites: TestSuite[]): Promise<TestReport> {
    const startTime = Date.now();
    this.pipelineController.start();
    
    const suiteResults: TestSuiteResult[] = [];
    let totalTests = 0;
    let passedTests = 0;
    let failedTests = 0;
    let skippedTests = 0;

    console.log(`🚀 Starting test orchestration with ${suites.length} suites`);
    console.log(`⚙️  Configuration: failFast=${this.config.failFast}, parallel=${this.config.parallel}`);

    for (const suite of suites) {
      console.log(`\n📋 Running test suite: ${suite.name}`);
      
      const suiteResult = await this.runSuite(suite);
      suiteResults.push(suiteResult);
      
      totalTests += suiteResult.tests.length;
      passedTests += suiteResult.tests.filter(t => t.success).length;
      failedTests += suiteResult.tests.filter(t => !t.success).length;
      skippedTests += suiteResult.tests.filter(t => t.errors && t.errors.some(e => e.includes('skipped'))).length;

      // Handle test failure with pipeline controller
      if (!suiteResult.success) {
        const failedTest = suiteResult.tests.find(t => !t.success);
        if (failedTest) {
          const shouldContinue = this.pipelineController.shouldContinue(
            failedTest.name,
            suite.name,
            (failedTest.errors || []).join('; ')
          );
          
          if (!shouldContinue) {
            console.log(`🛑 Pipeline stopped due to test failure`);
            console.log(`📋 Issue ID: ${failedTest.issueId || 'UNKNOWN'}`);
            console.log(`🔧 Remediation: ${failedTest.remediation || 'Check error details'}`);
            break;
          } else {
            // Attempt to handle the failure
            const retry = await this.pipelineController.handleFailure(
              failedTest.name,
              suite.name,
              (failedTest.errors || []).join('; ')
            );
            
            if (!retry) {
              console.log(`🛑 Pipeline stopped: max retries exceeded`);
              break;
            }
          }
        }
      }
    }

    const totalDuration = Date.now() - startTime;
    const successRate = totalTests > 0 ? (passedTests / totalTests) * 100 : 0;
    
    // Get pipeline result
    const pipelineResult = this.pipelineController.getResult();

    const report: TestReport = {
      timestamp: new Date().toISOString(),
      totalTests,
      passedTests,
      failedTests,
      skippedTests,
      totalDuration,
      suites: suiteResults,
      summary: {
        successRate,
        averageDuration: totalTests > 0 ? totalDuration / totalTests : 0,
        fastestTest: this.findFastestTest(suiteResults),
        slowestTest: this.findSlowestTest(suiteResults)
      }
    };

    await this.generateReport(report);
    return report;
  }

  private async runSuite(suite: TestSuite): Promise<TestSuiteResult> {
    const startTime = Date.now();
    const tests: TestResult[] = [];
    const errors: string[] = [];

    console.log(`  📝 Running ${suite.commands.length} commands`);

    if (this.config.parallel && suite.commands.length > 1) {
      // Run commands in parallel with concurrency limit
      const results = await this.runCommandsParallel(suite.commands, suite);
      tests.push(...results);
    } else {
      // Run commands sequentially
      for (const command of suite.commands) {
        try {
          console.log(`    🔧 Running: ${command.name}`);
          const result = await command.run(this.context);
          tests.push(result);
          
          if (result.success) {
            console.log(`    ✅ ${command.name} passed (${result.duration}ms)`);
          } else {
            console.log(`    ❌ ${command.name} failed: ${(result.errors || []).join(', ')}`);
            errors.push(...result.errors);
            
            // Fail fast if enabled
            if (suite.failFast !== false && this.config.failFast) {
              console.log(`    🛑 Stopping suite execution due to failure (failFast enabled)`);
              break;
            }
          }
        } catch (error) {
          const errorResult: TestResult = {
            name: command.name,
            success: false,
            duration: Date.now() - startTime,
            errors: [`Command execution failed: ${error}`],
            warnings: [],
            artifacts: [],
            metadata: {}
          };
          tests.push(errorResult);
          errors.push(`Command ${command.name} failed: ${error}`);
          
          if (suite.failFast !== false && this.config.failFast) {
            console.log(`    🛑 Stopping suite execution due to error (failFast enabled)`);
            break;
          }
        }
      }
    }

    const success = tests.every(t => t.success);
    const duration = Date.now() - startTime;

    console.log(`  📊 Suite "${suite.name}" completed: ${success ? 'PASSED' : 'FAILED'} (${duration}ms)`);

    return {
      name: suite.name,
      success,
      duration,
      tests,
      errors
    };
  }

  private async runCommandsParallel(commands: BaseCommand[], suite: TestSuite): Promise<TestResult[]> {
    const results: TestResult[] = [];
    const concurrency = Math.min(this.config.maxConcurrency, commands.length);
    
    console.log(`    🔄 Running ${commands.length} commands in parallel (max ${concurrency} concurrent)`);

    for (let i = 0; i < commands.length; i += concurrency) {
      const batch = commands.slice(i, i + concurrency);
      const batchPromises = batch.map(async (command) => {
        try {
          console.log(`      🔧 Running: ${command.name}`);
          const result = await command.run(this.context);
          
          if (result.success) {
            console.log(`      ✅ ${command.name} passed (${result.duration}ms)`);
          } else {
            console.log(`      ❌ ${command.name} failed: ${(result.errors || []).join(', ')}`);
          }
          
          return result;
        } catch (error) {
          console.log(`      ❌ ${command.name} error: ${error}`);
          return {
            name: command.name,
            success: false,
            duration: 0,
            errors: [`Command execution failed: ${error}`],
            warnings: [],
            artifacts: [],
            metadata: {}
          } as TestResult;
        }
      });

      const batchResults = await Promise.all(batchPromises);
      results.push(...batchResults);

      // Check for failures and fail fast if enabled
      const hasFailures = batchResults.some(r => !r.success);
      if (hasFailures && suite.failFast !== false && this.config.failFast) {
        console.log(`    🛑 Batch failed, stopping parallel execution (failFast enabled)`);
        break;
      }
    }

    return results;
  }

  private findFastestTest(suiteResults: TestSuiteResult[]): string {
    let fastest = '';
    let fastestTime = Infinity;

    for (const suite of suiteResults) {
      for (const test of suite.tests) {
        if (test.duration < fastestTime) {
          fastestTime = test.duration;
          fastest = test.name;
        }
      }
    }

    return fastest || 'N/A';
  }

  private findSlowestTest(suiteResults: TestSuiteResult[]): string {
    let slowest = '';
    let slowestTime = 0;

    for (const suite of suiteResults) {
      for (const test of suite.tests) {
        if (test.duration > slowestTime) {
          slowestTime = test.duration;
          slowest = test.name;
        }
      }
    }

    return slowest || 'N/A';
  }

  private async generateReport(report: TestReport): Promise<void> {
    const fs = await import('fs');
    const path = await import('path');
    
    try {
      const reportDir = path.dirname(this.config.reportPath);
      if (!fs.existsSync(reportDir)) {
        fs.mkdirSync(reportDir, { recursive: true });
      }

      fs.writeFileSync(this.config.reportPath, JSON.stringify(report, null, 2));
      console.log(`📄 Test report saved to: ${this.config.reportPath}`);
    } catch (error) {
      console.error(`❌ Failed to save test report: ${error}`);
    }
  }
}

File: backend/tests/orchestrator/TestOrchestrator.py
"""
TestOrchestrator - Orchestrates backend test execution with fail-fast behavior
"""

import asyncio
import json
import time
from typing import List, Dict, Any, Optional
from dataclasses import dataclass, asdict
from ..commands.BaseCommand import BaseCommand, TestContext, TestResult, TestStatus

@dataclass
class OrchestratorConfig:
    base_url: str
    timeout: int
    retries: int
    fail_fast: bool
    parallel: bool
    max_concurrency: int
    report_path: str
    database_url: Optional[str] = None
    api_key: Optional[str] = None

@dataclass
class TestSuite:
    name: str
    commands: List[BaseCommand]
    timeout: Optional[int] = None
    retries: Optional[int] = None
    fail_fast: Optional[bool] = None

@dataclass
class TestReport:
    timestamp: str
    total_tests: int
    passed_tests: int
    failed_tests: int
    skipped_tests: int
    total_duration: float
    suites: List[Dict[str, Any]]
    summary: Dict[str, Any]

class TestOrchestrator:
    """Orchestrates test execution with fail-fast behavior"""
    
    def __init__(self, config: OrchestratorConfig):
        self.config = config
        self.context = TestContext(
            base_url=config.base_url,
            timeout=config.timeout,
            retries=config.retries,
            fail_fast=config.fail_fast,
            database_url=config.database_url,
            api_key=config.api_key
        )
    
    async def run_suites(self, suites: List[TestSuite]) -> TestReport:
        """Run multiple test suites"""
        start_time = time.time()
        suite_results = []
        total_tests = 0
        passed_tests = 0
        failed_tests = 0
        skipped_tests = 0
        
        print(f"🚀 Starting test orchestration with {len(suites)} suites")
        print(f"⚙️  Configuration: failFast={self.config.fail_fast}, parallel={self.config.parallel}")
        
        for suite in suites:
            print(f"\n📋 Running test suite: {suite.name}")
            
            suite_result = await self.run_suite(suite)
            suite_results.append(suite_result)
            
            total_tests += len(suite_result["tests"])
            passed_tests += len([t for t in suite_result["tests"] if t["status"] == TestStatus.SUCCESS.value])
            failed_tests += len([t for t in suite_result["tests"] if t["status"] == TestStatus.FAILURE.value])
            skipped_tests += len([t for t in suite_result["tests"] if t["status"] == TestStatus.SKIPPED.value])
            
            # Fail fast if enabled and suite failed
            if self.config.fail_fast and not suite_result["success"]:
                print(f"❌ Suite '{suite.name}' failed, stopping execution (failFast enabled)")
                break
        
        total_duration = time.time() - start_time
        success_rate = (passed_tests / total_tests * 100) if total_tests > 0 else 0
        
        report = TestReport(
            timestamp=time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
            total_tests=total_tests,
            passed_tests=passed_tests,
            failed_tests=failed_tests,
            skipped_tests=skipped_tests,
            total_duration=total_duration,
            suites=suite_results,
            summary={
                "success_rate": success_rate,
                "average_duration": total_duration / total_tests if total_tests > 0 else 0,
                "fastest_test": self.find_fastest_test(suite_results),
                "slowest_test": self.find_slowest_test(suite_results)
            }
        )
        
        await self.generate_report(report)
        return report
    
    async def run_suite(self, suite: TestSuite) -> Dict[str, Any]:
        """Run a single test suite"""
        start_time = time.time()
        tests = []
        errors = []
        
        print(f"  📝 Running {len(suite.commands)} commands")
        
        if self.config.parallel and len(suite.commands) > 1:
            # Run commands in parallel with concurrency limit
            test_results = await self.run_commands_parallel(suite.commands, suite)
            tests.extend(test_results)
        else:
            # Run commands sequentially
            for command in suite.commands:
                try:
                    print(f"    🔧 Running: {command.name}")
                    result = await command.run(self.context)
                    tests.append(asdict(result))
                    
                    if result.status == TestStatus.SUCCESS:
                        print(f"    ✅ {command.name} passed ({result.duration:.2f}s)")
                    else:
                        print(f"    ❌ {command.name} failed: {', '.join(result.errors)}")
                        errors.extend(result.errors)
                        
                        # Fail fast if enabled
                        if suite.fail_fast is not False and self.config.fail_fast:
                            print(f"    🛑 Stopping suite execution due to failure (failFast enabled)")
                            break
                            
                except Exception as error:
                    error_result = TestResult(
                        name=command.name,
                        status=TestStatus.ERROR,
                        duration=time.time() - start_time,
                        errors=[f"Command execution failed: {error}"],
                        warnings=[],
                        artifacts=[],
                        metadata={}
                    )
                    tests.append(asdict(error_result))
                    errors.append(f"Command {command.name} failed: {error}")
                    
                    if suite.fail_fast is not False and self.config.fail_fast:
                        print(f"    🛑 Stopping suite execution due to error (failFast enabled)")
                        break
        
        success = all(t["status"] == TestStatus.SUCCESS.value for t in tests)
        duration = time.time() - start_time
        
        print(f"  📊 Suite '{suite.name}' completed: {'PASSED' if success else 'FAILED'} ({duration:.2f}s)")
        
        return {
            "name": suite.name,
            "success": success,
            "duration": duration,
            "tests": tests,
            "errors": errors
        }
    
    async def run_commands_parallel(self, commands: List[BaseCommand], suite: TestSuite) -> List[Dict[str, Any]]:
        """Run commands in parallel with concurrency limit"""
        results = []
        concurrency = min(self.config.max_concurrency, len(commands))
        
        print(f"    🔄 Running {len(commands)} commands in parallel (max {concurrency} concurrent)")
        
        for i in range(0, len(commands), concurrency):
            batch = commands[i:i + concurrency]
            batch_tasks = []
            
            for command in batch:
                task = asyncio.create_task(self.run_single_command(command))
                batch_tasks.append(task)
            
            batch_results = await asyncio.gather(*batch_tasks, return_exceptions=True)
            
            for j, result in enumerate(batch_results):
                if isinstance(result, Exception):
                    error_result = TestResult(
                        name=batch[j].name,
                        status=TestStatus.ERROR,
                        duration=0,
                        errors=[f"Command execution failed: {result}"],
                        warnings=[],
                        artifacts=[],
                        metadata={}
                    )
                    results.append(asdict(error_result))
                    print(f"      ❌ {batch[j].name} error: {result}")
                else:
                    results.append(asdict(result))
                    if result.status == TestStatus.SUCCESS:
                        print(f"      ✅ {result.name} passed ({result.duration:.2f}s)")
                    else:
                        print(f"      ❌ {result.name} failed: {', '.join(result.errors)}")
            
            # Check for failures and fail fast if enabled
            has_failures = any(r["status"] != TestStatus.SUCCESS.value for r in results[-len(batch):])
            if has_failures and suite.fail_fast is not False and self.config.fail_fast:
                print(f"    🛑 Batch failed, stopping parallel execution (failFast enabled)")
                break
        
        return results
    
    async def run_single_command(self, command: BaseCommand) -> TestResult:
        """Run a single command"""
        try:
            print(f"      🔧 Running: {command.name}")
            result = await command.run(self.context)
            return result
        except Exception as error:
            return TestResult(
                name=command.name,
                status=TestStatus.ERROR,
                duration=0,
                errors=[f"Command execution failed: {error}"],
                warnings=[],
                artifacts=[],
                metadata={}
            )
    
    def find_fastest_test(self, suite_results: List[Dict[str, Any]]) -> str:
        """Find the fastest test"""
        fastest = ""
        fastest_time = float('inf')
        
        for suite in suite_results:
            for test in suite["tests"]:
                if test["duration"] < fastest_time:
                    fastest_time = test["duration"]
                    fastest = test["name"]
        
        return fastest or "N/A"
    
    def find_slowest_test(self, suite_results: List[Dict[str, Any]]) -> str:
        """Find the slowest test"""
        slowest = ""
        slowest_time = 0
        
        for suite in suite_results:
            for test in suite["tests"]:
                if test["duration"] > slowest_time:
                    slowest_time = test["duration"]
                    slowest = test["name"]
        
        return slowest or "N/A"
    
    async def generate_report(self, report: TestReport) -> None:
        """Generate test report"""
        try:
            import os
            report_dir = os.path.dirname(self.config.report_path)
            if not os.path.exists(report_dir):
                os.makedirs(report_dir, exist_ok=True)
            
            with open(self.config.report_path, 'w') as f:
                json.dump(asdict(report), f, indent=2)
            
            print(f"📄 Test report saved to: {self.config.report_path}")
        except Exception as error:
            print(f"❌ Failed to save test report: {error}")
